source/common/http/filter/extauth.cc

```diff
@@ -0,0 +1,308 @@
+ * request is stalled until the auth request completes.
+ *
+ * If the auth request returns HTTP 200, the original request is allowed
+ * to continue. If any headers are listed in the extauth filter's "headers"
+ * array, those headers will be copied from the auth response into the
+ * original request (overwriting any duplicate headers).
+ *
+ * If the auth request returns anything other than HTTP 200, the original
+ * request is rejected. The full response from the auth service is returned
+ * as the response to the rejected request.
+ *
+ * Note that at present, a call to the external service is made for _every
+ * request_ being routed.
+ */
+
+static LowerCaseString header_to_add(std::string("x-ambassador-calltype"));
+static LowerCaseString value_to_add(std::string("extauth-request"));
+
+ExtAuth::ExtAuth(ExtAuthConfigConstSharedPtr config) : config_(config) {}
+
+ExtAuth::~ExtAuth() { ASSERT(!auth_request_); }
+
+/* dump headers for debugging */
+void ExtAuth::dumpHeaders(const char* what, HeaderMap* headers) {
+#ifndef NVLOG
+  ENVOY_STREAM_LOG(trace, "ExtAuth headers ({}):", *callbacks_, what);
+
+  headers->iterate(
+      [](const HeaderEntry& header, void* context) -> HeaderMap::Iterate {
+        ENVOY_STREAM_LOG(trace, "  '{}':'{}'", *static_cast<StreamDecoderFilterCallbacks*>(context),
+                         header.key().c_str(), header.value().c_str());
+        return HeaderMap::Iterate::Continue;
+      },
+      static_cast<void*>(callbacks_));
+#endif
+}
+
+/*
+ * decodeHeaders is called at the point that the HTTP machinery handling
+ * the request has parsed the HTTP headers for this request.
+ *
+ * Our primary job here is to construct the request to the auth service
+ * and start it executing, but we also have to be sure to save a pointer
+ * to the incoming request headers in case we need to modify them in
+ * flight.
+ */
+
+FilterHeadersStatus ExtAuth::decodeHeaders(HeaderMap& headers, bool) {
+  // Remember that we have _not_ finished talking to the auth service...
+
+  auth_complete_ = false;
+
+  // ...and hang onto a pointer to the original request headers.
+  //
+  // Note that using a reference here won't work. Move semantics are the
+  // root of this issue, I think.
+
+  request_headers_ = &headers;
+
+  // Debugging.
+  dumpHeaders("decodeHeaders", request_headers_);
+
+  // OK, time to get the auth-service request set up. Create a
+  // RequestMessageImpl to hold all the details, and start it off as a
+  // copy of the incoming request's headers.
+
+  MessagePtr reqmsg(new RequestMessageImpl(HeaderMapPtr{new HeaderMapImpl(headers)}));
+
+  // We do need to tweak a couple of things. To start with, has a change
+  // to the path we hand to the auth service been configured?
+
+  if (!config_->path_prefix_.empty()) {
+    // Yes, it has. Go ahead and prepend it to the reqmsg path.
+
+    std::string path = reqmsg->headers().insertPath().value().c_str();
```

Comment:
I agree that the getString is clearer, but It looks like that it will need to allocate.

getString():
```
std::string getString() const { return std::string(buffer_.ref_, string_length_); 
```

c_str():
```
const char* c_str() const { return buffer_.ref_; }
```

I'm not sure if `std::string(buffer_.ref_, string_length_)` gets optimized.

Reply:
It will. RVO copy elision is all but guaranteed in this case. In fact, in C++17 it will be *guaranteed*

Reply:
👍